README.md
# About
This project displays the current starry sky relative to the selected latitude using planispheres made by Dominic Ford. See https://in-the-sky.org/planisphere/index.php for more information.

HISTORY.md
#Tracking progress

added virtualenv
imageMaker:
  Planisphere.py:
    commented out LaTeX related code
    print english content only
    print pngs only
    moved imageMaker to different location
    grabbed latitude and ratio data
  holder.py:
    allowed background color
    removed bottom half
    moved circle for starwheel up
    printed point where starwheel needs to go in percentage relative to holder image
    extracted clockwheel locations
    made big white background
  starwheel.py:
    extracted jan1st in degrees location
    removed month ring from starwheel
  choose colors

planisphere:
  tested basic functionality
  added images into canvas
  added all images
  displaying right images
  added draw function
  starwheel lines up pretty close to correct position
  corrected starwheel alignment
  planisphere is working
  made viewing window bigger
  added color designs

script.js
const canvas = document.querySelector("#planisphereContainer");
const ctx = canvas.getContext("2d");
const dpi = window.devicePixelRatio;
let holderImg, starwheelImg, degreestoJanFirst, rotate, ratio;
const latitude = 50; // the latitude that the page initially displays
const enlarge = 1.3; // used to enlarge the viewing window of the holder

// fixes canvas blur
function fix_dpi() {
    //create a style object that returns width and height
    let style = {
        height() {
            return +getComputedStyle(canvas).getPropertyValue('height').slice(0, -2);
        },
        width() {
            return +getComputedStyle(canvas).getPropertyValue('width').slice(0, -2);
        }
    }
    //set the correct attributes for a crystal clear image!
    canvas.setAttribute('width', style.width() * dpi);
    canvas.setAttribute('height', style.height() * dpi);
};
fix_dpi();

async function main(latitude) {
  holderImg = new Image();
  starwheelImg = new Image();
  await loadImage(`./images/holder${latitude}.png`, holderImg);
  await loadImage(`./images/starwheel${latitude}.png`, starwheelImg);
  scaleToFit();
  draw();
}

fetch("./data.json")
  .then(response => {
    if (!response.ok) {
      throw new Error('Network response was not OK');
    }
    return response.json();
  })
  .then(async function(data) {
    const { latitudes } = data;
    ratio = data.ratio;
    degreestoJanFirst = data.degreestoJanFirst;
    addSelectInput(latitudes);
    await main(latitude);
    setInterval(draw, 600000); // draw every minute

    // fixes image when window is resized
    window.addEventListener("resize", function() {
      fix_dpi();
      scaleToFit();
      draw();
    });
  })
  .catch(error => {
    console.error('There has been a problem with your fetch operation:', error);
  });

helpers.js
function addSelectInput(latitudes) {
  const select = document.createElement("select");
  latitudes.forEach(latitude => {
    const option = document.createElement("option");
    option.setAttribute("value", latitude);
    option.innerHTML = `Latitude ${latitude}`;
    select.appendChild(option);
  });
  select.value = latitude;
  document.querySelector(".container").appendChild(select);

  select.addEventListener("change", function(e) {
    main(e.target.value);
  });
}
// from: https://riptutorial.com/html5-canvas/example/19169/scaling-image-to-fit-or-fill-
// scales the images to fit into canvas while preserving their aspect ratio
function scaleToFit(){
  // get the scale
  holderImg.canvasScale = Math.min(canvas.width / holderImg.width, canvas.height / holderImg.height) * enlarge;
  // get the top left position of the image
  holderImg.canvasX = (canvas.width / 2) - (holderImg.width / 2) * holderImg.canvasScale;
  holderImg.canvasY = (canvas.height / 2) - (holderImg.height / 2) * holderImg.canvasScale;

    // to make top of starwheel visible
  holderImg.offset = holderImg.width * holderImg.canvasScale / 10;
  const starwheelCenter = holderImg.canvasY + holderImg.height * holderImg.canvasScale * Math.abs(1 - ratio);
  starwheelImg.canvasX = holderImg.canvasX;
  starwheelImg.canvasY = starwheelCenter - holderImg.width / 2 * holderImg.canvasScale;
  holderImg.canvasScale;
}
function loadImage(url, imgObj) {
  return new Promise((res, rej) => {
    imgObj.addEventListener('load', function() {
      res();
    }, false);
    imgObj.src = url; // Set source path
  });
}
// gets the degree to rotate the star wheel to.
function getRotation(latitude, degreestoJanFirst) {
  // get local time
  const now = new Date();
  let hour = now.getHours();
  // from: https://stackoverflow.com/questions/8619879/javascript-calculate-the-day-of-the-year-1-366
  const start = new Date(now.getFullYear(), 0, 0);
  const diff = (now - start) + ((start.getTimezoneOffset() - now.getTimezoneOffset()) * 60 * 1000);
  const oneDay = 1000 * 60 * 60 * 24;
  let day = Math.floor(diff / oneDay) - 1;
  const degPerDay = 360 / 365;
  const degPerHour = 360 / 24;
  let janFirst, currentHourDeg, deg, currentDayDeg;
  if (latitude >= 0) { // north of equator
    janFirst = degreestoJanFirst.north;
    hour = 24 - hour;
    currentHourDeg = (hour * degPerHour + 270) % 360 ;
    currentDayDeg = (janFirst + degPerDay * day) % 360;
  } else { // south of equator
    janFirst = degreestoJanFirst.south;
    currentHourDeg = (hour * degPerHour + 270) % 360;
    day = 365 - day;
    currentDayDeg = (janFirst + degPerDay * day) % 360;
  }
  if (currentDayDeg > currentHourDeg) {
    deg = 360 - currentDayDeg + currentHourDeg;
  } else {
    deg = currentHourDeg - currentDayDeg;
  }
  return deg;
}
function draw() {
  ctx.clearRect(0, 0, canvas.width, canvas.height);

  let rotate = getRotation(latitude, degreestoJanFirst);
  const x = holderImg.canvasX + (holderImg.width * holderImg.canvasScale) / 2;
  const y = starwheelImg.canvasY + holderImg.offset + (holderImg.width * holderImg.canvasScale) / 2;
  // Matrix transformation
  ctx.translate(x, y);
  ctx.rotate(rotate * Math.PI / 180);
  ctx.translate(-x, -y);
  // rotate starwheel
  ctx.drawImage(starwheelImg, holderImg.canvasX, starwheelImg.canvasY + holderImg.offset, holderImg.width * holderImg.canvasScale, holderImg.width * holderImg.canvasScale);

  // Matrix transformation
  ctx.translate(x, y);
  ctx.rotate(-rotate * Math.PI / 180);
  ctx.translate(-x, -y);

  ctx.drawImage(holderImg, holderImg.canvasX, holderImg.canvasY + holderImg.offset, holderImg.width * holderImg.canvasScale, holderImg.height * holderImg.canvasScale);  
}

imageMaker/planisphere.py
#!/usr/bin/python3
# planisphere.py
# -*- coding: utf-8 -*-
#
# The python script in this file makes the various parts of a model planisphere.
#
# Copyright (C) 2014-2022 Dominic Ford <dcf21-www@dcford.org.uk>
#
# This code is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# You should have received a copy of the GNU General Public License along with
# this file; if not, write to the Free Software Foundation, Inc., 51 Franklin
# Street, Fifth Floor, Boston, MA  02110-1301, USA

# ----------------------------------------------------------------------------

"""
This is the top level script for drawing all the parts needed to build planispheres which work at a range of
different latitudes. They are rendered in PDF, SVG and PNG image formats.

Additionally, we use LaTeX to build a summary document for each latitude, which includes all of the parts needed
to build a planisphere for that latitude, and instructions as to how to put them together.
"""

import os
import subprocess
import time

import text
from alt_az import AltAzGrid
from holder import Holder
from settings import fetch_command_line_arguments
from starwheel import StarWheel

# Create output directory
os.system("rm -Rf output")
os.system("mkdir -p output/planispheres")

arguments = fetch_command_line_arguments()
theme = arguments['theme']

# Render planisphere in all available languages
# for language in text.text:
# Fernando edit
language = "en"

# Render climates for latitudes at 5-degree spacings from 10 deg -- 85 deg, plus 52N
for latitude in list(range(-80, 90, 5)) + [52]:

    # Do not make equatorial planispheres, as they don't really work
    if -10 < latitude < 10:
        continue

    # Boolean flag for which hemiphere we're in
    southern = latitude < 0

    # A dictionary of common substitutions
    # subs = {
    #     "dir_parts": "output/planisphere_parts",
    #     "dir_out": "output/planispheres",
    #     "abs_lat": abs(latitude),
    #     "ns": "S" if southern else "N",
    #     "lang": language,
    #     "lang_short": "" if language == "en" else "_{}".format(language)
    # }

    # settings = {
    #     'language': language,
    #     'latitude': latitude,
    #     'theme': theme
    # }

    # Fernando edit render individual parts here. Gather from bottom of scripts
    # Render the star wheel for the planisphere
    StarWheel(settings={
        'latitude': latitude,
        'language': 'en',
        'theme': arguments['theme'],
    }).render_to_file(
        filename=f"./output/planispheres/starwheel{latitude}",
        img_format=arguments['img_format'],

    )
        # Render the holder for the planisphere
    Holder(settings={
        'latitude': latitude,
        'language': 'en'
    }).render_to_file(
        filename=f"./output/planispheres/holder{latitude}",
        img_format=arguments['img_format']
    )

    # Render the various parts of the planisphere
    # StarWheel(settings=settings).render_all_formats(
    #     filename="{dir_parts}/starwheel_{abs_lat:02d}{ns}_{lang}".format(**subs)
    # )

    # Holder(settings=settings).render_all_formats(
    #     filename="{dir_parts}/holder_{abs_lat:02d}{ns}_{lang}".format(**subs)
    # )

    # AltAzGrid(settings=settings).render_all_formats(
    #     filename="{dir_parts}/alt_az_grid_{abs_lat:02d}{ns}_{lang}".format(**subs)
    # )

    # Copy the PDF versions of the components of this astrolabe into LaTeX's working directory, to produce a
    # PDF file containing all the parts of this astrolabe
    # os.system("mkdir -p doc/tmp")
    # os.system("cp {dir_parts}/starwheel_{abs_lat:02d}{ns}_{lang}.pdf doc/tmp/starwheel.pdf".format(**subs))
    # os.system("cp {dir_parts}/holder_{abs_lat:02d}{ns}_{lang}.pdf doc/tmp/holder.pdf".format(**subs))
    # os.system("cp {dir_parts}/alt_az_grid_{abs_lat:02d}{ns}_{lang}.pdf doc/tmp/altaz.pdf".format(**subs))

    # with open("doc/tmp/lat.tex", "wt") as f:
    #     f.write(r"${abs_lat:d}^\circ${ns}".format(**subs))

    # Wait for cairo to wake up and close the files
    time.sleep(1)

    # Build LaTeX documentation
    # for build_pass in range(3):
    #     subprocess.check_output("cd doc ; pdflatex planisphere{lang_short}.tex".format(**subs), shell=True)

    # os.system("mv doc/planisphere{lang_short}.pdf "
    #           "{dir_out}/planisphere_{abs_lat:02d}{ns}_{lang}.pdf".format(**subs))

    # For the English language planisphere, create a symlink with no language suffix in the filename
    # if language == "en":
    #     os.system("ln -s planisphere_{abs_lat:02d}{ns}_en.pdf "
    #               "{dir_out}/planisphere_{abs_lat:02d}{ns}.pdf".format(**subs))

    # Clean up the rubbish that LaTeX leaves behind
    # os.system("cd doc ; rm -f *.aux *.log *.dvi *.ps *.pdf")


imageMaker/main_planisphere.sh
#!/bin/bash
# main_planisphere.sh
# -*- coding: utf-8 -*-
#
# The bash script in this file makes the various parts of a model planisphere.
#
# Copyright (C) 2014-2022 Dominic Ford <dcf21-www@dcford.org.uk>
#
# This code is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# You should have received a copy of the GNU General Public License along with
# this file; if not, write to the Free Software Foundation, Inc., 51 Franklin
# Street, Fifth Floor, Boston, MA  02110-1301, USA

# ----------------------------------------------------------------------------

# Delete any previous output from running this script
rm -Rf __pycache__ *.pyc output
rm -Rf doc/*.aux doc/*.dvi doc/*.log doc/*.pdf doc/*.ps doc/tmp doc/*.out

# Run the python 3 script which generates planisphere models for a wide range of latitudes
python3 ./planisphere.py --format png 

# Clean up temporary files which get made along the way
rm -Rf __pycache__ *.pyc
rm -Rf doc/*.aux doc/*.dvi doc/*.log doc/*.pdf doc/*.ps doc/tmp doc/*.out


imageMaker/themes.py
# theme.py
# -*- coding: utf-8 -*-
#
# The python script in this file defines the available themes for the planisphere.
#
# Copyright (C) 2019 André Werlang <dcf21-www@dcford.org.uk>
#
# This code is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# You should have received a copy of the GNU General Public License along with
# this file; if not, write to the Free Software Foundation, Inc., 51 Franklin
# Street, Fifth Floor, Boston, MA  02110-1301, USA

# ----------------------------------------------------------------------------

themes = {
    "default":
        {
            # "background": (1, 1, 1, 0),
            # "date": (0, 0, 0, 1),
            # "edge": (0, 0, 0, 1),
            # "shading": (0.9, 0.9, 0.9, 1),
            # "grid": (0.75, 0.75, 0.75, 1),
            # "stick": (0.25, 0.25, 0.25, 1),
            # "star": (0, 0, 0, 1),
            # "constellation": (0, 0, 0, 1)
            "background": (1, 1, 1, 0),
            "date": (0, 0, 0, 1),
            "edge": (0, 0, 0, 1),
            "shading": (0.9, 0.9, 0.9, 1),
            "grid": (0.75, 0.75, 0.75, 1),
            "stick": (0.25, 0.25, 0.25, 1),
            "star": (0, 0, 0, 1),
            "constellation": (0, 0, 0, 1)
        },
    "dark":
        {
            "background": (0.2, 0.25, 0.45, 1),
            "date": (1, 1, 1, 1),
            "edge": (0.45, 0.45, 0.45, 1),
            "shading": (0.1, 0.1, 0.1, 1),
            "grid": (0.3, 0.3, 0.3, 1),
            "stick": (0.28, 0.35, 0.55, 1),
            "star": (1, 1, 1, 1),
            "constellation": (0.6, 0.5, 0.65, 1)
        }
}
